### DISCUSSION TITLE
Allow run cmd to clone private repo

### DISCUSSION THREAD

**Discussion by Bascy** [community/community#26584] [Actions]
<p>I’m tryng out Github actions to build buld our many repositories created for PlatformIO IoT development.<br>
Many of our repo’s have dependencies on other private repos, and those dependent repo’s are cloned as part of the platformIO build process.</p>
<p>So my action contains a <code>run</code> statement that will try to clone other private repo’s but this isnt working.<br>
The clone action ends in a fatal error:</p>
<blockquote>
<p>git version 2.31.1<br>
(https: //github. com/ I-Connect/csMainV5/runs/2385097024?check_suite_focus=true#step:7:125)Cloning into ‘/home/runner/work/csMainV5/csMainV5/.pio/.cache/tmp/pkg-installing-ojc7shvw’…<br>
(<a href="https://github" rel="noopener nofollow ugc">https://github</a>. com/I-Connect/csMainV5/runs/2385097024?check_suite_focus=true#step:7:126)fatal: could not read Username for ‘https:// github. com/)’: No such device or address</p>
</blockquote>
<p>I’ve tried the suggestion with an AccesToken on my account, and added it to the repo being build but that doesnt seem to work.</p>
<p>How can I get this to work?</p>
<pre><code class="lang-auto">name: PlatformIO CI
on: [push]
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        token: ${{ secrets.ACCESS_TOKEN }}

    - name: Cache pip
      uses: actions/cache@v2
      with:
        path: ~/.cache/pip
        key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
        restore-keys: |
          ${{ runner.os }}-pip-

    - name: Cache PlatformIO
      uses: actions/cache@v2
      with:
        path: ~/.platformio
        key: ${{ runner.os }}-${{ hashFiles('**/lockfiles') }}

    - name: Set up Python
      uses: actions/setup-python@v2
    - name: Install PlatformIO
      run: |
        python -m pip install --upgrade pip
        pip install --upgrade gitpython
        pip install --upgrade platformio

    - name: Run PlatformIO
      run: pio run -e debug -e release
</code></pre>

**Comment by airtower-luna** [comment]
<p><code>actions/checkout</code> configures the token only for the local repository, not for any the build might be retrieving. The easiest approach would probably be to copy the <code>http.https://github.com/.extraheader</code> setting from the repository to the user level config. Make sure to delete it again after checking out what you need!</p>

**Comment by Bascy** [comment]
<p>Thank you for your reply. I don’t quite understand what you mean with the <code>http.https://github.com/.extraheader</code> setting but I’m not sure something like that will work…</p>
<p>The project that is being build has a config file stating its dependencies and that is used by the build process  (<code>pio run ...</code>) to clone the necessary repositories. I don’t want to repeat those dependencies in the action script.</p>
<p>I’m just reading about adding a access-token to the repo url … As the dependencies are specified as git urls, it might work if I add an access token there. Will try that tonight  … hmm but that would mean having explicit credentials in my source code - don’t want that</p>

**Comment by airtower-luna** [comment]
<p>If you look at the log of a workflow using <code>actions/checkout</code> you can see something like this:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/0/3/03d447451fd145e1e189911c036195fbedafcdcf.png" data-download-href="/uploads/short-url/xSeGbVjWh3OiLg0h4PuE8wzC1h.png?dl=1" title="grafik"><img src="https://user-images.githubusercontent.com/5814919/181098576-8ac58899-a9e0-44ac-83ce-4df54ca357bb.png" alt="grafik" data-base62-sha1="xSeGbVjWh3OiLg0h4PuE8wzC1h" width="690" height="239" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/0/3/03d447451fd145e1e189911c036195fbedafcdcf_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">grafik</span><span class="informations">757×263 39.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>The last command in that list adds an <code>Authorization</code> header to HTTPS requests to <code>github.com</code>, with a value based on the token, <em>for the local repository</em> (<code>--local</code>). You could use a similar approach to set up authorization before you fetch the other repositories, probably at the user level.</p>

**Comment by Bascy** [comment, marked as answer]
<p>I tried copying from local to global but that didnt go very well, then I found this:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/marketplace/actions/setup-git-credentials" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:400/400;"><img src="https://avatars.githubusercontent.com/u/45044662?s=400&amp;amp;v=4" class="thumbnail" width="400" height="400"></div>

<h3><a href="https://github.com/marketplace/actions/setup-git-credentials" target="_blank" rel="noopener nofollow ugc">setup-git-credentials - GitHub Marketplace</a></h3>


  <p><span class="label1">Allow cloning private repositories</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>It’s exactly what I needed</p>
